javalib/src/main/scala/java/util/regex/Matcher.scala

```diff
@@ -46,287 +45,410 @@ object Matcher {
       qrLoop(s, idx, sb).toString
     }
   }
+
 }
 
 final class Matcher private[regex] (var _pattern: Pattern,
-                                    var inputSequence: CharSequence)
+                                    var _inputSequence: CharSequence)
     extends MatchResult {
 
-  private def withRE2Regex[A](f: RE2RegExpOps => A): A =
-    _pattern.withRE2Regex(f)
+  private var currentPattern = _pattern
 
   private var hasMatch  = false
   private var hasGroups = false
   private var appendPos = 0
 
-  private var groups = withRE2Regex { re2op =>
-    val regex = re2op.ptr
-    Array.ofDim[(Int, Int)](cre2.numCapturingGroups(regex) + 1)
+  private var _regionStart = 0
+  private var _regionEnd   = _inputSequence.length
+
+  private var _groupCount = currentPattern.compiled.groupCount()
+
+  // This default differs from JVM but describes actual SN behavior.
+  // That behavior is a bug, awaiting a fix.
+  private var anchoringBoundsInUse = false
+
+  private def clearGroup0(g: Array[Int]) = {
+    g(0) = -1
+    g(1) = -1
   }
 
-  private var lastAnchor: Option[cre2.anchor_t] = None
+  private def createGroups(nGroups: Int): Array[Int] = {
 
-  private[regex] def inputLength = inputSequence.length
+    // +1 is for groups(0). It holds full match, but is not part of groupCount.
+    // "groups" is an old-style 2-D row-major array, where Array(index) holds
+    // start for group number index and Array(index + 1) holds the
+    // corresponding end. In SN, 1-D arrays are much more efficient, albeit
+    // much less elegant.
```

Comment:
1D array are only more *size* efficient for large constants such as unicode tables. Otherwise there is no reason not to use 2D arrays for dynamically-sized data if it's easier to understand.

Reply:
Entirely gone in j.u.r.Matcher.scala in 2019-04-16 code with improved delegation.

I verified that re2s.Matcher.scala uses 1-D arrays.